libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1718,11 +1718,22 @@ void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator, Type co
 			m_context << u256(2) << Instruction::EXP << Instruction::MUL;
 		break;
 	case Token::SAR:
-		// NOTE: SAR rounds differently than SDIV
-		if (m_context.evmVersion().hasBitwiseShifting() && !c_valueSigned)
-			m_context << Instruction::SHR;
+		if (m_context.evmVersion().hasBitwiseShifting())
+			m_context << Instruction::SAR;
 		else
-			m_context << u256(2) << Instruction::EXP << Instruction::SWAP1 << (c_valueSigned ? Instruction::SDIV : Instruction::DIV);
+		{
+			if (c_valueSigned)
+				m_context.appendInlineAssembly(R"({
+					let xor_mask := sub(0,slt(value_to_shift,0))
```

Comment:
Please add a space after each comma :)

Reply:
Can we assume the value to be clean, or do we have to call `signextend` before the check? Since we previously only used `SHR`, it should be fine, I think.

Reply:
I think it should be fine, but I can try to find some test cases for that and/or read through the code to see whether the assumption is valid. I think I'll also add a few more test cases for more odd and even, signed and unsigned values of different bit widths in general.

Reply:
Added some test cases that verify that ``signextend`` is not needed.